api/src/main/java/com/xwiki/googleapps/internal/GoogleAppsEventListener.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.xwiki.googleapps.internal;
+
+import com.xpn.xwiki.doc.XWikiDocument;
+import org.xwiki.bridge.event.ApplicationReadyEvent;
+import org.xwiki.bridge.event.DocumentUpdatedEvent;
+import org.xwiki.component.phase.InitializationException;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.observation.EventListener;
+import org.xwiki.observation.event.Event;
+
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Registered object to listen to document changes.
+ *
+ * @version $Id$
+ * @since 3.0
+ */
+class GoogleAppsEventListener implements EventListener
```

Comment:
Where / how is this event listener used?

Reply:
Ouch. That's a good. That got moved away for readability and to prevent double-instantiation but was not constructed when the manager starts. Now fixed in 70368b23ea6498a229ef3f4e26fb78dbfaffce12 .

Reply:
Still, why isn't this a component?

Reply:
I've spent quite some time trying to make the various objects a component and, at the end, considered it a failure. Here is the reason:

- There is no practice (apparently) to make package-internal components. This is a requirement for almost all classes here: They make no sense if not configured with the rest and many methods need to be public even though their function is only internal (this follows from attempts but also from a discussion with Vincent Massol).
- When building, e.g. GoogleAppsXWikiObjects, you need to refer to GoogleAppsManager and conversely. Same for just about any component. That means you need providers... not very elegant and bringing a big uncertainty as to whether your object is actually built.
- Normal practice for components is to apply "the interface dance": Each component's function is described in an interface and a component is an implementation of this interface. This works but doubles the code which is really not useful for functions that internal.

I agree that having injection for some of the objects (in particular loggers) would be a lot better but requiring this for all seems to be too much and expose too much of the API.

Reply:
> There is no practice (apparently) to make package-internal components.

There **is** a practice and I believe I mentioned it to you.. We do have **internal** components that don't implement any interface, where you can have package protected methods or public internal methods as you like. See https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-extension/xwiki-platform-extension-handlers/xwiki-platform-extension-handler-xar/src/main/java/org/xwiki/extension/xar/internal/doc/InstalledExtensionDocumentTree.java for instance.

> That means you need providers...

I don't see the problem with that.

> Normal practice for components is to apply "the interface dance"

Again. You can have **internal** components that don't have an interface. The best practice requires an interface only for public components.

Reply:
Well, that example:

- is a public class
- has a public constructor
- ... and has public methods

While the last is probably avoidable, the two first are necessary for a component to be built.
So that makes the class part of the public API even if without public methods.

paul

Reply:
Yes, a component class needs to be public and its default constructor needs to be public as well, but that doesn't make it public API.

> So that makes the class part of the public API even if without public methods.

No, what makes it a public API is the package. Anything inside the ``*.internal.*`` package is **internal API**. Anything not inside the ``*.internal.*`` package is public API. Of course, you can use any public class, but using internal API means your code can **break** at any upgrade (we can add, remove, modify internal APIs as we wish without caring how they are used outside, because they are internal API). For public APIs we need to maintain backwards compatibility. So is that your only concern regarding using components: the fact that they "become public API"? Because they don't.

Reply:
So I can remove any annotation of Unstable in there too?

This issue is what I remember now to have been the only problem I met.
I'll try it again (it might be that I was dissisatisfied by using provider everywhere too but I can accept that).
I'll report when done.

Reply:
``@Unstable`` annotation doesn't make sense inside the **``internal``** package, for internal APIs.

Reply:
I have now pushed a version with just components and no complex constructors anymore.
It would be cool to have the ability to have package-internal classes... but that's just a wish for the future.
Thanks for insisting. It does make the code more focussed (e.g. the manager does not need to grab everything).
paul

Reply:
Hello @mflorea,
I believe that a review for the latest "componentization" (commit d8109496a41c3042e419161da0acfd15b56fe3c6 ) has it complete now.
I would love a review.
paul